replication/binlogsyncer.go

```diff
@@ -729,27 +734,18 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
 		b.nextPos.Pos = uint32(event.Position)
 		log.Infof("rotate to %s", b.nextPos)
 	case *GTIDEvent:
-		if b.gset == nil {
-			break
-		}
 		u, _ := uuid.FromBytes(event.SID)
-		err := b.gset.Update(fmt.Sprintf("%s:%d", u.String(), event.GNO))
-		if err != nil {
-			return errors.Trace(err)
-		}
+		currentGTID = fmt.Sprintf("%s:%d", u.String(), event.GNO)
 	case *MariadbGTIDEvent:
-		if b.gset == nil {
-			break
-		}
 		GTID := event.GTID
-		err := b.gset.Update(fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
-		if err != nil {
-			return errors.Trace(err)
-		}
+		currentGTID = fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber)
 	case *XIDEvent:
-		event.GSet = b.getGtidSet()
+		event.GSet, err = b.updateGtidSet(currentGTID)
 	case *QueryEvent:
-		event.GSet = b.getGtidSet()
+		event.GSet, err = b.updateGtidSet(currentGTID)
```

Comment:
does this still have a problem with the `Query` event Begin in DML event sequences like following
```
| bin.000001 | 16597 | Gtid           |       101 |       16662 | SET @@SESSION.GTID_NEXT= '9afe121c-40c2-11e9-9ec7-0242ac110002:55'                                                                                                                                                                                                                                                                                                                                        |
| bin.000001 | 16662 | Query          |       101 |       16730 | BEGIN                                                                                                                                                                                                                                                                                                                                                                                                     |
| bin.000001 | 16730 | Table_map      |       101 |       16791 | table_id: 11993 (sharding1.t2)                                                                                                                                                                                                                                                                                                                                                                            |
| bin.000001 | 16791 | Update_rows    |       101 |       17277 | table_id: 11993 flags: STMT_END_F                                                                                                                                                                                                                                                                                                                                                                         |
| bin.000001 | 17277 | Xid            |       101 |       17308 | COMMIT /* xid=51083 */
```